sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TypeNamesShouldNotMatchNamespaces.cs

```diff
@@ -0,0 +1,99 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class TypeNamesShouldNotMatchNamespaces : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4041";
+        private const string MessageFormat = "Change the name of that type to be different from an existing namespace.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        // Based on https://msdn.microsoft.com/en-us/library/gg145045%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396
+        private static ISet<string> frameworkNamespaces =
+            new HashSet<string>
+            {
+                "accessibility", "activities", "addin", "build", "codedom", "collections",
+                "componentmodel", "configuration", "csharp", "custommarshalers", "data",
+                "dataflow", "deployment", "device", "diagnostics", "directoryservices",
+                "drawing", "dynamic", "enterpriseservices", "globalization", "identitymodel",
+                "interopservices", "io", "jscript", "linq", "location", "management", "media",
+                "messaging", "microsoft", "net", "numerics", "printing", "reflection", "resources",
+                "runtime", "security", "server", "servicemodel", "serviceprocess", "speech",
+                "sqlserver", "system", "tasks", "text", "threading", "timers", "transactions",
+                "uiautomationclientsideproviders", "visualbasic", "visualc", "web", "win32",
+                "windows", "workflow", "xaml", "xamlgeneratednamespace", "xml"
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var declaration = (BaseTypeDeclarationSyntax)c.Node;
+                var symbolDeclaredccess = c.SemanticModel.GetDeclaredSymbol(declaration)?.DeclaredAccessibility;
+
+                ReportIfNameClashesWithFrameworkNamespace(declaration?.Identifier, symbolDeclaredccess, c);
+            },
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.EnumDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var declaration = (DelegateDeclarationSyntax)c.Node;
+                var symbolDeclaredccess = c.SemanticModel.GetDeclaredSymbol(declaration)?.DeclaredAccessibility;
+
+                ReportIfNameClashesWithFrameworkNamespace(declaration?.Identifier, symbolDeclaredccess, c);
+            },
+            SyntaxKind.DelegateDeclaration);
```

Comment:
This block is exactly the same as the previous one except from the cast. I am pretty sure you could use `c.SemanticModel.GetDeclaredSymbol(c.Node)` directly.